ilc/client/navigationEvents/setupEvents.js

```diff
@@ -1,13 +1,84 @@
+let navigationHook;
+let currentUrl = window.location.href;
+
+const capturedEventListeners = {
+    hashchange: [],
+    popstate: [],
+};
+
+const routingEventsListeningTo = [
+    'hashchange',
+    'popstate'
+];
+
+const addEventListener = window.addEventListener;
+const removeEventListener = window.removeEventListener;
+
+function callCapturedEventListeners(eventArguments) {
+    if (!eventArguments) {
+        return;
+    }
+
+    const eventType = eventArguments[0].type;
+
+    if (!(routingEventsListeningTo.indexOf(eventType) >= 0)) {
+        return;
+    }
+
+    capturedEventListeners[eventType].forEach((listener) => {
+        try {
+            listener.apply(this, eventArguments);
+        } catch (error) {
+            setTimeout(() => {
+                throw error;
+            });
+        }
+    });
+}
+
+function patchedAddEventListener(eventName, fn) {
+    if (
+        typeof fn === "function" &&
+        routingEventsListeningTo.includes(eventName) &&
+        !capturedEventListeners[eventName].includes(fn)
+    ) {
+        capturedEventListeners[eventName].push(fn);
+        return;
+    }
+
+    return addEventListener.apply(this, arguments);
+}
+
+function patchedRemoveEventListener(eventName, fn) {
+    if (typeof fn === "function" && routingEventsListeningTo.includes(eventName)) {
+        capturedEventListeners[eventName] = capturedEventListeners[eventName].filter((listener) => listener !== fn);
+        return;
+    }
+
+    return removeEventListener.apply(this, arguments);
+}
+
 function fireRoutingEvent() {
     window.dispatchEvent(new CustomEvent('ilc:before-routing'));
 }
 
+function setupNavigationHook(hook) {
```

Comment:
Could we have multiple hooks here?

Reply:
Updated, I have added the ability to do it